### ISSUE TITLE
[feature] Automatic fee bumping

### ISSUE THREAD

**Issue by bitromortac** [spesmilo/electrum#7310]
I'm currently thinking about which functionality would be needed for anchor commitments (#6588). It seems like one would need an _active_ method that ensures timely confirmation of commitment transactions via CPFP, finding a compromise between fees and confirmation time. This could also be useful for a users' transactions, where the user would submit a confirmation target to the auto-fee-bumper that ensures the confirmation within that timeframe given a budget it has available. Was there some discussion about something like that?

**Comment by SomberNight** [comment]
The issue with implementing generic automatic fee bump is that it needs to access to private keys, and those are usually not readily available.
To bump the fee of a tx, we need to re-sign it; and to sign it we need private keys. If the user has a software wallet with a password, we need to prompt the user to enter their password; or if they have a hardware wallet, prompt them to connect it/etc.

**Comment by bitromortac** [comment]
True, here I was thinking about a very specific application namely for confirming anchor commitment transactions. There the Lightning wallet always has access to the private keys to sign. I just thought that we could reuse this functionality and make it available to the user as well, to ~guarantee transaction confirmation within some height. This could become useful when the mempool becomes really full. Security could be improved by giving a budget for fee bumping and then reserving UTXOs to meet the budget. We could cache the private keys for these UTXOs, but it would only work for software wallets.

Issues @ecdsa mentioned are that this could lead to reinforcing behavior (if everybody does it it could create a race to the fee moon :)), but I think user-set budgets limit this. Another point is that one is exposed to manipulation by the server(s?), by wrong mempool data. Is mempool data gathered from several sources or from a single server?

**Comment by SomberNight** [comment]
> I was thinking about a very specific application namely for confirming anchor commitment transactions. There the Lightning wallet always has access to the private keys to sign. I just thought that we could reuse this functionality and make it available to the user as well, to ~guarantee transaction confirmation within some height. This could become useful when the mempool becomes really full. Security could be improved by giving a budget for fee bumping and then reserving UTXOs to meet the budget. We could cache the private keys for these UTXOs, but it would only work for software wallets.

Sure, we will need this for LN anchor commitment txs.
Re doing something similar for regular user-facing onchain txs, so potentially generalising the fee bumping behaviour, let's just think the use case through:
- most Electrum users probably do not run the application for a long time, they fire it up to do some stuff and then close it
- this means any kind of fee bumping, including for LN, would only work when it so happens that the user has the wallet open
- note that "caching private keys in memory" or similar, does not really work in this use case, as the user might have already closed and reopened the wallet since originally signing/broadcasting the tx
- alternatively, we could just keep private keys unencrypted in memory (every time the wallet is opened, prompt user for password, and either decrypt keys and keep them in memory, or keep password in memory)

I really would rather not store the privkeys/password in memory (though note that in the special case of the Android app, the password is already kept in memory).
An alternative would be to prompt the user for the password every time the software wants to bump the fee, but that might be a bit annoying and not great UX.

Another approach we could use instead is pre-sign multiple transactions at different fee levels, store them in the wallet file, and broadcast them over time.

In any case, I think this is not that important. Specifically, if you are thinking we could implement this first and then implement anchors, and adapt this to be used by anchors, then my hunch is that would risk wasting some effort/work. Rather, we should implement anchors before this, implement fee bumping as part of anchors or after anchors, and then later potentially expand fee bumping to more general use.

**Comment by bitromortac** [comment]
Yes, I agree and don't want to complicate things, just wanted to see if something like this could be useful after anchors.

**Comment by chrissv** [comment]
I am a recent "victim" of a commitment transaction having a too-low fee. I force-closed a channel to a now-defunct node, and the last channel activity (apparently a while ago) resulted in an updated commitment transaction being signed with 9.1 sat/vByte. So when that transaction was broadcast as part of the FC, it is languishing in the mempool. Since I don't own either the input or output addresses, I can't RBF or CPFP. If there were a dust-level output to one of my wallet addresses included in the transaction, I could at least CPFP that transaction to get it processed.

I'm not sure what to do now since I don't have any hope of the transaction being confirmed. Very disappointing (and expensive).

**Comment by SomberNight** [comment]
This functionality has been added in the meantime, as part of the new txbatcher module:
https://github.com/spesmilo/electrum/blob/2ea4874600179799683dbb970a02039bb892da97/electrum/txbatcher.py#L364
